### PR TITLE
Fix X-Forwarded-* headers in openvsx-proxy

### PR DESCRIPTION
**Pull request by jeanp413**
## Description

Fix X-Forwarded-* headers in openvsx-proxy, [openvsx will read these values](https://github.com/gitpod-io/openvsx/blob/8624398257663b06e56e90caefee4aaa6385da70/server/src/main/java/org/eclipse/openvsx/util/UrlUtil.java#L151-L163) to properly create links to extension resources 

## How to test

1. Open workspace 
2. Do some queries in vscode extension view and check response in dev tools links should point to openvsx-proxy
3. Check extensions install sucessfully

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/openvsx-proxy/pkg/run.go
Comment (akosyakov) [inline components/openvsx-proxy/pkg/run.go:172]: What if these headers already present ie someone using Gitpod behind another proxy. Should not we preserve them then?

Comment (felladrin) [inline components/openvsx-proxy/pkg/run.go:172]: Not sure if it's needed for this case, but maybe you also want to set "X-Forwarded-Port", as some customers were asking for it on [another occasion](https://github.com/gitpod-io/gitpod/pull/11253).

- https://github.com/gitpod-io/gitpod/pull/11253

Comment (jeanp413) [reply components/openvsx-proxy/pkg/run.go]: `X-Forwarded-Port` is not relevant for this case

Comment (jeanp413) [reply components/openvsx-proxy/pkg/run.go]: Added logic to preserve headers if they are present but from logs the headers are already present when the request reaches openvsx-proxy in the preview environment so there's already a proxy in front :smile: 
```
{"func":"request_handler","level":"info","message":"handling request","request":"\"POST /vscode/gallery/extensionquery HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: application/json;api-version=3.0-preview.1\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nCache-Control: no-cache\\r\\nContent-Length: 236\\r\\nContent-Type: application/json\\r\\nOrigin: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nPragma: no-cache\\r\\nReferer: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com/\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: empty\\r\\nSec-Fetch-Mode: cors\\r\\nSec-Fetch-Site: same-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nVscode-Sessionid: 37f5a423-2e86-4705-af53-fa9b9c3c461e1660244402281\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\nX-Market-Client-Id: VSCode 1.70.0\\r\\nX-Market-User-Id: 70064d9c-b88e-4cdc-bb46-a2d0e7f5d4cb\\r\\n\\r\\n\"","request_content_length":"236","request_id":"060cc98b-7dfb-4e55-a3db-d1e699ec36a5","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":947558,"duration_human":"947.558µs","func":"request_handler","level":"info","message":"processing request finished","request":"POST /vscode/gallery/extensionquery 236 653ddf16318e0343fd552c6dccda25c78e550d3d","request_content_length":"236","request_id":"060cc98b-7dfb-4e55-a3db-d1e699ec36a5","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"request_handler","level":"info","message":"handling request","request":"\"POST /vscode/gallery/extensionquery HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: application/json;api-version=3.0-preview.1\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nCache-Control: no-cache\\r\\nContent-Length: 236\\r\\nContent-Type: application/json\\r\\nOrigin: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nPragma: no-cache\\r\\nReferer: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com/\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: empty\\r\\nSec-Fetch-Mode: cors\\r\\nSec-Fetch-Site: same-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nVscode-Sessionid: 37f5a423-2e86-4705-af53-fa9b9c3c461e1660244402281\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\nX-Market-Client-Id: VSCode 1.70.0\\r\\nX-Market-User-Id: 70064d9c-b88e-4cdc-bb46-a2d0e7f5d4cb\\r\\n\\r\\n\"","request_content_length":"236","request_id":"f95d2aca-12d7-40fa-95e1-84e60d220a7e","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":583004,"duration_human":"583.004µs","func":"request_handler","level":"info","message":"processing request finished","request":"POST /vscode/gallery/extensionquery 236 653ddf16318e0343fd552c6dccda25c78e550d3d","request_content_length":"236","request_id":"f95d2aca-12d7-40fa-95e1-84e60d220a7e","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":204211504,"duration_human":"204.211504ms","func":"transport_roundtrip","level":"info","message":"upstream call finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"response_handler","level":"info","message":"handling response","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"func":"response_handler","level":"info","message":"successfully stored response to cache","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"duration":1184627,"duration_human":"1.184627ms","func":"response_handler","level":"info","message":"processing response finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"duration":209098713,"duration_human":"209.098713ms","func":"request_handler","hit_cache_backup":true,"hit_cache_regular":false,"level":"info","message":"request finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_content_length":"0","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"request_handler","level":"info","message":"handling request","request":"\"GET /vscode/asset/wesleyegberto/vscode-java-tests/1.5.0/Microsoft.VisualStudio.Services.Icons.Default HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: image\\r\\nSec-Fetch-Mode: no-cors\\r\\nSec-Fetch-Site: cross-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\n\\r\\n\"","request_content_length":"0","request_id":"30ecf423-2fd5-45bb-884a-91cd107bab87","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
```

File: components/openvsx-proxy/pkg/handler.go
Comment (andreafalzetti) [inline components/openvsx-proxy/pkg/handler.go:44]:
```suggestion
```

👀

**Review by andreafalzetti — APPROVED** [review]
Tested and works as advertised. 

/hold in case of anyone with more context on this has major comments

File: components/openvsx-proxy/pkg/handler.go
Comment (jeanp413) [reply components/openvsx-proxy/pkg/handler.go:44]: yep I'll remove it before merging :+1:

**Review by felladrin — APPROVED** [review]

**Comment by akosyakov** [comment]
@jeanp413 I noticed that now if i navigate to Marketplace link it takes out to our proxy, not open-vsx.org as on main:
Screenshot 2022-08-12 at 07 58 19
 
The link on main as well points to proxy but it resolves to upstrea:
Screenshot 2022-08-12 at 07 59 33

Did we change anything that could break it?

File: components/openvsx-proxy/pkg/handler.go
Comment (akosyakov) [inline components/openvsx-proxy/pkg/handler.go:52]: it seems to break: https://github.com/gitpod-io/gitpod/pull/12071#issuecomment-1212756291

Do we really need to change it? OpenVSX should care about `X-Forwarded*` to create links or it also looks at Host header for it?

Comment (akosyakov) [reply components/openvsx-proxy/pkg/handler.go:52]: It is interesting though in context of Self Hosted in air gapped scenario where maybe a user machine does not have access to open-vsx.org directly but only via our proxy. In this case it would be even desirable. Not sure. cc @corneliusludmann

**Review by akosyakov — COMMENTED** [review]
great work, but we need to agree on https://github.com/gitpod-io/gitpod/pull/12071#issuecomment-1212756291 before landing

also cleaning up commit history :)

**Comment by jeanp413** [comment]
> @jeanp413 I noticed that now if i navigate to Marketplace link it takes out to our proxy, not open-vsx.org as on main:

@akosyakov  That's expected behavior, as the openvsx api will use the `x-forwarded-*` headers, I can add a hardcoded check in the reverse-proxy that if the endpoint is `/vscode/item` then delete the `x-forwarded-*` headers so openvsx use the prod url in this case `open-vsx.org`

**Comment by akosyakov** [comment]
> @akosyakov That's expected behavior, as the openvsx api will use the x-forwarded-* headers, I can add a hardcoded check in the reverse-proxy that if the endpoint is /vscode/item then delete the x-forwarded-* headers so openvsx use the prod url in this case open-vsx.org

actually let's go with it, I think it is minor, and can be rather beneficial for self hosted installation.

Feel free to clean up commits and unhold 🙏

**Pull request merged**
